### PR TITLE
Add support for exl2-quantized models

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

Add support for exl2 quantization

Mostly straightforward, changes to existing code:

* Wrap quantizer parameters in a small wrapper to avoid passing around untyped tuples and needing to repack them as a dict.
* Move scratch space computation to warmup, because we need the maximum input sequence length to avoid allocating huge scratch buffers that OOM.

**Draft:** needs a rebase, exllama kernels seem non-deterministic, so logprobs sometimes change slightly?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/utils/weights.py [hunk 1]
```diff
@@ -9,6 +10,93 @@
...
+
+        w = Exl2Weight(**tensors)
+        w.q_scale_max /= 256
+        w.q_perm = w.q_perm.short()
+        w.q_invperm = w.q_invperm.short()
+
+        return w
+
+    @staticmethod
+    def make_group_map(q_groups: torch.Tensor, num_qrows: int):
```
Comment (Narsil) [inline server/text_generation_server/utils/weights.py:73]: This function actually already exists in `exllamav2.py`  let's not rewrite it.

Comment (Narsil) [reply server/text_generation_server/utils/weights.py:73]: I had a few comments about it, but since it's coming from external repo (which should be indicated somewhere). It's not worth changing anything, the benefit of leaving as-is is that we can just pull updates whenever they come.

**Review by Narsil — COMMENTED** [review]
Things seem to be working.

In general though, I feel like the PR is currently adding way too many indirections than necessary, every logic should be in their appropriate files and never add more.

- `layers/tensor_parallel`  is about the actual parallel logic, it can have some slight variations based on quantization, but only to change higher order loading logic.
- `weights.py` is all about creating the actual individual tensors required on the model. This one knows about quantization and how to shard tensors.
- `layers/gptq` anything else that is GPTQ specific, not about loading/sharding tensors and more about running init phase (scratch buffers) and actual forwards.

File: docs/source/basic_tutorials/launcher.md [hunk 2]
```diff
@@ -62,6 +62,7 @@ Options:
           Possible values:
           - awq:              4 bit quantization. Requires a specific AWQ quantized model: <https://hf.co/models?search=awq>. Should replace GPTQ models wherever possible because of the better latency
           - eetq:             8 bit quantization, doesn't require specific model. Should be a drop-in replacement to bitsandbytes with much better performance. Kernels are from <https://github.com/NetEase-FuXi/EETQ.git>
+          - exl2:             Variable bit quantization. Requires a specific EXL2 quantized model: <https://hf.co/models?search=exl2>. Requires exllama2 kernels and does support tensor parallelism (num_shard > 1)
```
Comment (Narsil) [inline docs/source/basic_tutorials/launcher.md:65]:
```suggestion
          - exl2:             Variable bit quantization. Requires a specific EXL2 quantized model: <https://hf.co/models?search=exl2>. Requires exllama2 kernels and does NOT support tensor parallelism (num_shard > 1)
```

Comment (Narsil) [reply docs/source/basic_tutorials/launcher.md:65]: ? :D

File: docs/source/basic_tutorials/launcher.md [hunk 3]
```diff
@@ -62,6 +62,7 @@ Options:
           Possible values:
           - awq:              4 bit quantization. Requires a specific AWQ quantized model: <https://hf.co/models?search=awq>. Should replace GPTQ models wherever possible because of the better latency
           - eetq:             8 bit quantization, doesn't require specific model. Should be a drop-in replacement to bitsandbytes with much better performance. Kernels are from <https://github.com/NetEase-FuXi/EETQ.git>
+          - exl2:             Variable bit quantization. Requires a specific EXL2 quantized model: <https://hf.co/models?search=gptq>. Requires exllama2 kernels
```
[the remark this answers is not in the document]
Comment (danieldk) [reply docs/source/basic_tutorials/launcher.md]: Fixed.

File: integration-tests/models/test_flash_gemma_exl2.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@pytest.mark.asyncio
+@pytest.mark.private
+async def test_flash_gemma_exl2(flash_gemma_exl2, generous_response_snapshot):
+    response = await flash_gemma_exl2.generate(
+        "Test request", max_new_tokens=10, decoder_input_details=True
+    )
+
+    assert response.details.generated_tokens == 10
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/models/test_flash_gemma_exl2.py]: Done.

File: integration-tests/models/test_flash_llama_exl2.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import pytest
+
+
+@pytest.fixture(scope="module")
+def flash_llama_exl2_handle(launcher):
+    with launcher(
+        "turboderp/Llama-3-8B-Instruct-exl2",
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/models/test_flash_llama_exl2.py:7]: Removed the Gemma tests.

File: launcher/src/main.rs [hunk 6]
```diff
@@ -55,6 +55,9 @@ enum Quantization {
     /// Should be a drop-in replacement to bitsandbytes with much better performance.
     /// Kernels are from <https://github.com/NetEase-FuXi/EETQ.git>
     Eetq,
+    /// Variable bit quantization. Requires a specific EXL2 quantized model:
+    /// <https://hf.co/models?search=gptq>. Requires exllama2 kernels.
```
[the remark this answers is not in the document]
Comment (danieldk) [reply launcher/src/main.rs]: Fixed.

[the remark this answers is not in the document]
Comment (danieldk) [reply launcher/src/main.rs]: Added.

File: server/text_generation_server/layers/gptq/exllamav2.py [hunk 7]
```diff
@@ -1,5 +1,7 @@
 # Adapted from turboderp exllama: https://github.com/turboderp/exllamav2
 
+from typing import Dict, Optional, Type, Union
+from text_generation_server.utils.weights import Exl2Weight, GPTQWeight
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/gptq/exllamav2.py]: I moved Exl2Weight to `layers.exl2` (`layers/gptq/exllamav2.py` would be ideal, but it's a conditional import).

File: server/text_generation_server/models/__init__.py [hunk 8]
```diff
@@ -881,6 +886,8 @@ def get_model(
         raise NotImplementedError("4bit quantization is not supported for AutoModel")
     elif quantize == "eetq":
         raise NotImplementedError("Eetq quantization is not supported for AutoModel")
+    if quantize == "exl2":
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/__init__.py]: Fixed.

File: server/text_generation_server/models/custom_modeling/flash_mistral_modeling.py [hunk 9]
```diff
@@ -119,8 +119,15 @@ def _load_gqa(config, prefix: str, weights):
     assert config.hidden_size % config.num_attention_heads == 0
     assert config.num_attention_heads % weights.process_group.size() == 0
 
+    prefixes = [f"{prefix}.q_proj", f"{prefix}.k_proj", f"{prefix}.v_proj"]
+
+    if config.quantize == "exl2":
+        return TensorParallelColumnLinear.load_multi(
+            config, prefixes=prefixes, dim=0, weights=weights, bias=False
+        )
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/models/custom_modeling/flash_mistral_modeling.py]: I switched it completely over to `TensorParallelColumnLinear.load_multi`.

File: server/text_generation_server/server.py [hunk 10]
```diff
@@ -89,7 +89,7 @@ async def FilterBatch(self, request, context):
         return generate_pb2.FilterBatchResponse(batch=filtered_batch.to_pb())
 
     async def Warmup(self, request, context):
-        if self.quantize == "gptq":
+        if self.quantize in ["exl2", "gptq"]:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/server.py]: Changed.

File: server/text_generation_server/utils/weights.py [hunk 11]
```diff
@@ -9,6 +10,93 @@
...
+            tensors["q_groups"], tensors["q_weight"].shape[0]
+        )
+        tensors["q_perm"] = torch.argsort(tensors["q_invperm"]).to(torch.int)
+
+        w = Exl2Weight(**tensors)
+        w.q_scale_max /= 256
+        w.q_perm = w.q_perm.short()
+        w.q_invperm = w.q_invperm.short()
+
+        return w
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: I have moved the loading to the relevant `Weights` methods. The ancillary (`q_perm`, `q_group_map`) tensors are now created in `layers/gptq/exllamav2.py`.

File: server/text_generation_server/utils/weights.py [hunk 12]
```diff
@@ -9,6 +10,93 @@
...
+                group_map += [rows - j]
+
+        return torch.tensor(group_map, dtype=torch.short, device=q_groups.device)
+
+    @property
+    def device(self) -> torch.device:
+        return self.q_weight.device
+
+    @staticmethod
+    def _load_tensor_keys() -> Set[str]:
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: Removed.

File: server/text_generation_server/utils/weights.py [hunk 13]
```diff
@@ -9,6 +10,93 @@
...
+            rows = qrows * 32 // bits
+            for j in range(rows):
+                group_map += [i]
+                group_map += [rows - j]
+
+        return torch.tensor(group_map, dtype=torch.short, device=q_groups.device)
+
+    @property
+    def device(self) -> torch.device:
+        return self.q_weight.device
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: The issue is that in GPTQ the weights are called `qweight` and in exl2 `q_weight`, so the `device` property is there to get the device independent of the tensor/field names. We could also rename the `q_weight` field in `Exl2Weight` to `qweight` if you prefer.

File: server/text_generation_server/utils/weights.py [hunk 14]
```diff
@@ -379,11 +503,22 @@ def get_multi_weights_row(self, prefix: str, quantize: str):
...
+                bits=bits,
+                groupsize=groupsize,
+                use_exllama=use_exllama,
+            )
         else:
             weight = self.get_sharded(f"{prefix}.weight", dim=1)
         return weight
 
+    def _get_exl2_bits(self) -> int:
+        return int(self.gptq_bits)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: Not used, so I removed it.

File: server/text_generation_server/utils/weights.py [hunk 15]
```diff
@@ -9,6 +10,93 @@
...
+                )
+
+        tensors["q_group_map"] = Exl2Weight.make_group_map(
+            tensors["q_groups"], tensors["q_weight"].shape[0]
+        )
+        tensors["q_perm"] = torch.argsort(tensors["q_invperm"]).to(torch.int)
+
+        w = Exl2Weight(**tensors)
+        w.q_scale_max /= 256
+        w.q_perm = w.q_perm.short()
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: Moved this to a proper dataclass constructor + __post_init__.

File: server/text_generation_server/utils/weights.py (hunk 1 above)
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: Moved back to `exllamav2.py`

File: server/text_generation_server/layers/gptq/exllamav2.py [hunk 16]
```diff
@@ -135,6 +106,13 @@ def set_device(device):
 
 def create_exllama_buffers(max_total_tokens: int):
     global FIXED_BYTES, LAYERS, DEVICE
+
+    # Find the size of the scratch space.
+    for layer in LAYERS:
+        FIXED_BYTES = max(
+            FIXED_BYTES, layer.scratch_space_fixed(max_input_len=max_total_tokens)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/gptq/exllamav2.py]: :picardfacepalm:

File: docs/source/basic_tutorials/launcher.md (hunk 2 above)
Comment (danieldk) [reply docs/source/basic_tutorials/launcher.md:65]: :picardfacepalm:

File: server/text_generation_server/layers/gptq/exllamav2.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/gptq/exllamav2.py]: https://github.com/turboderp/exllamav2/blob/8a57be1edfd0c3e6387a14876a359ef95f477739/exllamav2/model.py#L477

We need to know to know the max memory use of all layers to allocate the scratch buffer, before we can pass (slices of) the scratch buffer to the layers.

[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/gptq/exllamav2.py]: Did replace the loop with a more readable for-comprehension now.

File: server/text_generation_server/utils/weights.py [hunk 17]
```diff
@@ -9,6 +10,93 @@
...
+                tensors[key] = weight.get_tensor(name)
+            except RuntimeError:
+                raise RuntimeError(
+                    f"Cannot load `exl2`-quantized weight {name}, make sure the model is already quantized."
+                )
+
+        tensors["q_group_map"] = Exl2Weight.make_group_map(
+            tensors["q_groups"], tensors["q_weight"].shape[0]
+        )
+        tensors["q_perm"] = torch.argsort(tensors["q_invperm"]).to(torch.int)
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: Did that.

File: server/text_generation_server/layers/tensor_parallel.py [hunk 18]
```diff
@@ -108,22 +141,39 @@ def load_qkv(cls, config, prefix: str, weights, bias: bool):
...
-            prefixes, quantize=config.quantize, dim=dim
-        )
-
+        weight = weights.get_weights_col(prefix, config.quantize)
         if bias:
-            b = [weights.get_sharded(f"{p}.bias", dim=0) for p in prefixes]
-            bias = torch.cat(b, dim=dim)
+            if config.quantize == "exl2":
+                # TODO: exl2 sharding.
+                bias = weights.get_tensor(f"{prefix}.bias")
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/layers/tensor_parallel.py]: Indeed, fixed.

File: integration-tests/conftest.py [hunk 19]
```diff
@@ -223,6 +229,10 @@ class GenerousResponseComparator(ResponseComparator):
     rtol = 0.75
 
 
+class IgnoreLogProbResponseComparator(ResponseComparator):
+    atol = math.inf
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/conftest.py]: Replaced by a bool `ignore_logprob`.

File: server/text_generation_server/utils/weights.py [hunk 20]
```diff
@@ -9,6 +10,93 @@
...
+
+        w = Exl2Weight(**tensors)
+        w.q_scale_max /= 256
```
[the remark this answers is not in the document]
Comment (danieldk) [reply server/text_generation_server/utils/weights.py]: https://github.com/huggingface/text-generation-inference/blob/cbced7f0f9ca0b62216223859b82a2632d1c7a1f/server/text_generation_server/layers/gptq/exllamav2.py#L60

File: integration-tests/models/test_flash_gemma_exl2.py [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@pytest.fixture(scope="module")
+def flash_gemma_exl2_handle(launcher):
+    with launcher(
+        "turboderp/Gemma-7B-it-exl2",
+        revision="3.0bpw",
+        # Set max input length to avoid OOM due to extremely large
+        # scratch buffer.
+        max_input_length=1024,
```
[the remark this answers is not in the document]
Comment (danieldk) [reply integration-tests/models/test_flash_gemma_exl2.py]: I fixed that it does not unconditionally use a 4096 input length, but the value given by warmup. The default is probably still too much. E.g., in Llama2 8B for the output layer

4096 length * 16 batch size 128,256 pieces * 2 sizeof(float16) = 15.7GiB

The scratch buffer is allocated for the worst-case.

File: server/text_generation_server/utils/weights.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply server/text_generation_server/utils/weights.py]: > The issue is that in GPTQ the weights are called qweight and in exl2 q_weight, 

Ouch. Didn't even see. I get it then.
